### PR TITLE
Added a simple IO monad

### PR DESCRIPTION
**Pull request by jannickj**
I couldn't find the IO monad anywhere, maybe I've misunderstood your API :p, anyway here is my simple implementation, the IOStream type needs some work but changing it won't have an effect on existing code, so its safe to use as is.

### CODE REVIEW

**Comment by mausch** [comment]
Hi @jannickj , could you explain a bit what use cases do you see for IO ? The IO monad is intentionally missing in FSharpx because it's not very useful in an impure language like F#. For one, you can't statically ensure that functions without IO are pure as in Haskell. And if you want to compose actions as first-class values you can just use `Async<'T>`.

**Comment by jannickj** [comment]
it has many uses, for one it tells the programmer that this program contains side effects. But a more important one is the use that  you can say
let! b = try blah blah
or in this case 
let! b = attempt blah blah
I personally think that just because the language is impure doesn't mean we so just give up purity in our code, see this example to see how to turn a impure type into a pure one
https://github.com/jannickj/reddit-steam-bot/blob/master/RedditSteamBot/RedditSteamBot/WatiN.fs

**Comment by forki** [comment]
we have iteratees for IO.

**Comment by mausch** [comment]
You can already use `Async<'T>` for all of that, and it's built-in and well supported. For example, IIRC it implements trampolining to avoid stack overflows.

**Comment by jannickj** [comment]
Yes I thought about Async, but for instance in my case my program was using a browser which is only supported to run one at the time. So my API would be lying if I used Async.

@forki Okay funny name for a type that is for IO actions but as long as there can be agreement on one.

**Comment by jannickj** [comment]
Okay I looked over iteratee and it seemes to me that it is only the reading / input part of the IO monad, and It seems more like it is designed for one purpose in mind. whereas the Haskell IO monad is meant for all side effect actions

**Comment by mausch** [comment]
> I think if you want to make it async you should have to use a monad converter from IO to async

[Haskell runs IO asynchronously](http://www.reddit.com/r/haskell/comments/18d81l/ghc_head_new_parallel_io_manager_merged/c8dvvzc). IMHO that's up to the interpreter more than the type itself. `Async<'T>` being a monad means that you can lift any arbitrary value to it, there is no special requirement that it has to do async I/O so there is no lie there.

**Comment by mausch** [comment]
BTW Iteratees come from http://okmij.org/ftp/Streams.html ... FSharpx didn't make up the name :-)

**Comment by jannickj** [comment]
@mausch hmm good point. but how would something like putStr and getChar work, because unlike the haskell IO async itsn't tied to a stream

**Comment by mausch** [comment]
`let getChar = async { return System.Console.Read() }`

**Comment by jannickj** [comment]
Okay you've convinced me, while I don't like the idea that we are using async( as it makes it look like you are doing an async action instead of an IO), it does make sense to me to use the core features of F#. Especially as these are so close.

I would however like to add the attempt function (try in haskell)

**Comment by mausch** [comment]
F# already has that built-in as well:

```haskell
try :: Exception e => IO a -> IO (Either e a)
```

```fsharp
static member Catch : Async<'T> -> Async<Choice<'T,exn>>
```

https://msdn.microsoft.com/en-us/library/ee353899.aspx

**Comment by jannickj** [comment]
Ah nice I totally missed when I went through it, which was weird to me at the time as I was certain I had seen it.  Oh well thanks for taking the time, even if it didn't lead to anything productive I atleast got my confidence in my code back :D

File: src/FSharpx.Core/ComputationExpressions/Monad.fs [hunk 1]
```diff
@@ -61,6 +61,12 @@ module Async =
     /// Right-to-left Kleisli composition
     let inline (<=<) x = flip (>=>) x
 
+    let attempt m = async { 
+                             try  let! outp = m
+                                  return Choice1Of2 <| outp
+                             with exn -> return Choice2Of2 exn
+                    }
+
```
Comment (panesofglass) [inline src/FSharpx.Core/ComputationExpressions/Monad.fs]: This function should be moved to [`FSharpx.Async`](https://github.com/fsprojects/FSharpx.Async) as an extension to `Async`.

Comment (mausch) [inline src/FSharpx.Core/ComputationExpressions/Monad.fs]: @panesofglass  that's fine, it's a duplicate function anyway, see https://github.com/fsprojects/fsharpx/pull/320#issuecomment-75843774

**Comment by mausch** [comment]
@jannickj I think this discussion was certainly interesting! Please close the issue if you have no further questions.

**Pull request closed without merging**